### PR TITLE
fix: set the text alignment to .Natural so it will be presented according to the language alignment rules (RTL support)

### PR DESCRIPTION
**Pull request by pichirichi**
Since iOS 9 the default value for text alignment is .Natural
In iOS 8 you must set the value ignorer to get the correct alignment when changing languages.
Originally Labels in iOS 8 didn't have the .Natural alignment value.

### CODE REVIEW

**Comment by gergelyorosz** [comment]
Nice change! Would you mind adding a unit test for this change as well, that verifies this value is set after initialization? You can follow the naming of the tests as the existing ones in `SkyFloatingLabelTextFieldTests.swift`

**Comment by pichirichi** [comment]
I don't understand how a new test case cause a previous test case to fail :-/

**Comment by gergelyorosz** [comment]
Let me give a hand with this. The tests look good, nice job with those! I'll see if I can repro what's going on...

**Comment by gergelyorosz** [comment]
I just checked and the `test_whenSettingInitialTextAligment_thenTextFieldTextAligmentIsNatural` test case fails on the build machine. Locally the test passed on iOS 9.2 (with XCode 7.2), however the build machine is running iOS9.3 (with XCode 7.3). 

I'm getting this installed locally and will check why the test fails - perhaps the default value is different on iOS 9.3? I'm not sure.

Once we can repro the test failing locally, it should be straightforward to fix it.

**Comment by intonarumori** [comment]
Why don't we just not set the text alignment explicitly and rely on the system?

**Comment by pichirichi** [comment]
iOS 9 does it but iOS 8 is not doing this setting, we can add an if to do this setting only for iOS 8 but I thought there is no harm in changing it for all iOS versions.

**Comment by intonarumori** [comment]
I think it's a better idea to just let iOS decide the default for us, and don't explicitly change it in our implementation. This way we will always be consistent with the rest of iOS.

If you want to get a specific alignment you can always do that in your own setup code, or in a subclass, but in general we shouldn't add this to the base implementation.

To me it seems that if we do this we will just get the proper alignment out of the box, without adding any logic.

**Comment by pichirichi** [comment]
If you don't do that you can mark our the RTL support from the pod. If you need to set extra configuration to support RTL the pod is not RTL compliance. I have an workaround that I can do on my project but other users won't know how to do that. It is up to you, you can reject the change.

**Comment by intonarumori** [comment]
My idea here is to keep additional logic to a minimum while following iOS behaviour as close as possible.

Right now I'm looking into removing all explicit alignment settings. This way we end up with the same behaviour that you would have with the basic `UILabel`, `UITextField` classes.

Do you have to do workarounds with those classes?
- if not, then my idea should work, maybe there's a bug somewhere, but I would rather fix that than to add explicit setters
- if yes, then this class is no different than the rest of iOS, so whoever works with RTL would probably know about what to do since they have to do it already with the base classes.

If you could provide a sample code of what you are trying to do and your workaround, I could understand it better, thanks.

**Comment by pichirichi** [comment]
I have several open radars on RTL and I'm in direct touch with the Apple team working on RTL.
Lets face it, iOS 8 bugs won't be fixed and the RTL behaviour is not correct on UILabels. 
.Natural value for UILabels was introduced just in iOS 9 (until then was only available for AttributedLabels).
We can limit the alignment to iOS 8 only. Will this work for you?

**Comment by gergelyorosz** [comment]
If iOS8 default RTL support has a bug, I think it's valid for us to fix it (so people won't have to do this). The point of using this control would be to not have to do this bugfixing for others using this control - may that be iOS8 issues or not.

If we can limit te alignment setting to iOS8, I would be happy with this 👍  - as it would also make it easier to drop this config, once we remove iOS8 support later down the line.

Thanks @pichirichi for being super helpful with this one!

**Comment by intonarumori** [comment]
@pichirichi thanks for the additional info.
Could you please link those radars, so we can add it here for future reference?

In the meantime I'm testing in iOS8 simulators and check the alignments.
If there are indeed iOS8 bugs, then yes let's add the fixes and link to the radars for reference in the code.

**Comment by pichirichi** [comment]
This is the relevant radar for this fix:
- [18334130] Plain text UILabel are not adjusted correctly for RTL languages (Right To Left)
closed/duplicate/fixed for iOS9

**Comment by intonarumori** [comment]
@pichirichi thanks, unfortunately I cannot find it on radar, maybe it was deleted since.
Is there a Stackoverflow or similar answer about the problem and the fix that we can include?

**Comment by pichirichi** [comment]
1. http://stackoverflow.com/questions/18744447/autolayout-rtl-uilabel-text-alignment
2. http://iphonemithun.blogspot.co.il/2014/03/how-to-set-alignment-from-rtl-or-ltr-of.html

**Comment by intonarumori** [comment]
I did couple of tests with the iOS8.3 simulators and I see multiple issues with our RTL implementation:

- setting `UILabel` text alignment to `Natural` doesn't have an effect on iOS8.3.
Based on this and the radar issue the title label is always left aligned, the fix in this PR probably has no effect on the title label.
```
print(label.textAlignment.rawValue) // 0
label.textAlignment = .Natural
print(label.textAlignment.rawValue) // 0
```

- I set the simulator language to Arabic, but our input alignment was still Left-to-Right, so our current alignment setting is probably wrong:
simulator screen shot 17 may 2016 15 29 22

- Then I disabled the content of `updateTextAligment` and set `self.textAlignment = .Natural`.
In this case the alignment seemed to be good for the input.
simulator screen shot 17 may 2016 15 29 45


To summarise:
- `self.textAlignment = .Natural` is sensible, we can use it for iOS<9.0
- we should revisit our current RTL implementation, especially `updateTextAligment`
- `titleLabel.textAlignment = .Natural`, in my tests this didn't have any effect, so we should investigate it more.

**Comment by pichirichi** [comment]
the test implementation has no Arabic support so it won't change the UI to RTL. Change the device to Hebrew and you'll see the change. iOS change App UI for RTL only if the application indicates it support the language.

**Comment by intonarumori** [comment]
Ok, I'm starting to get confused.
Mostly because a simple `UITextField` with text alignment set to `Natural` _did_ change to RTL.

**Comment by pichirichi** [comment]
RTL is not that straight forward (no pun intended) for those who speak/write LTR languages so I understand the confusion. I speak Hebrew so it was easier to adjust our App to RTL and identify bugs and problems with iOS and this is the experience I brought when I've suggested the addition to your pod.

This is an Hebrew screen capture on how the pod sample code work in Hebrew:
img_0124

**Comment by intonarumori** [comment]
There are multiple use-cases for a `UITextField` to be RTL:
- LTR layout, RTL language (ex. using Arabic with the example project)
- RTL layout, RTL language (ex. using Hebrew with the example project)

- In both cases iOS would use RTL for the textfield, but in our current implementation on `master` we determine the direction by using `UIApplication.sharedApplication().userInterfaceLayoutDirection == .LeftToRight` and setting the alignment explicitly. We should rework this.

- I tested your PR code, and `titleLabel.textAlignment = .Natural` has no effect on iOS8.3.
You can test it after the setter it will still be `.Left`. We can remove this line.

- 'self.textAlignment = .Natural' is sensible, please add the condition for iOS8 and move it to `init_SkyFloatingLabelTextField`.

**Comment by pichirichi** [comment]
There are two issues I'm trying to solve with this change:
1. how the title look on the field when you start to edit text.
2. how the text input behave (cursor direction) when you enter text.

This is what happen on iOS 8 when only  `self.textAlignment = .Natural ` is set.

screen shot 2016-05-17 at 6 00 26 pm

This is what happen on iOS 8 when only `self.titleLabel.textAlignment = .Natural` is set.
screen shot 2016-05-17 at 6 02 05 pm

and this is what happen when both are set to .Natural:
screen shot 2016-05-17 at 6 03 30 pm

**Comment by intonarumori** [comment]
@pichirichi you are 15 commits behind `master` right now.
I think you will see the same once you are up to date.

**Comment by pichirichi** [comment]
You are right that I was few commits behind and now I see the problem:
```
    private func updateTextAligment() {
        if(self.isLTRLanguage) {
            self.textAlignment = .Left
        } else {
            self.textAlignment = .Right
        }
    }

```
This block of text is not correct, to support multilingual you use .Natural and you never force Left/Right according to language unless you expect a specific alignment for all languages (you can look on WWDC session regarding RTL support [(link)](https://developer.apple.com/library/ios/documentation/MacOSX/Conceptual/BPInternational/SupportingRight-To-LeftLanguages/SupportingRight-To-LeftLanguages.html)). When you set the alignment to .Natural iOS will handle it out of the box without any change in the code or checking the user interface orientation. This block wasn't part of my RTL updates commit.

Also you need to remember the following:

- iOS 8 and iOS 9 don't behave the same with their RTL support.
- Changing language via project schema doesn't the exact behaviour like changing the device language.

**Comment by gergelyorosz** [comment]
thanks for going to this depth for both of you.

I'm happy to fix this issue - together with the other reported RTL issue as well, and perhaps @pichirichi you could review this fix, to make sure the changes I did are correct? I can issue a PR later today - at least I'll also get my head around RTL.

**Comment by k0nserv** [comment]
Closing due to being outdated. I will review the current RTL implementation as per the above
